LibPDF: Fix memory leak, make it more usable

BenWiederhake opened the pull request:
This PR addresses some problems in LibPDF:
- Fix a potential memory leak at the expense of slightly higher memory consumption.
- Multiple headers couldn't stand alone due to missing other headers. That is, trying to compile a CPP file with just `#include <LibPDF/Something.h>` in it used to fail, now it works.
- Break up two dependency cycles (Object←→Value, and Document←→Parser)
- Fix a comment with wrong math in it

Some of these issues were found by just trying to compile a CPP file that only contains an `#include`. I'll try to make that tool public eventually, in a different PR.

alimpfard commented on Userland/Libraries/LibPDF/Value.h:
```diff
@@ -179,9 +178,9 @@ class Value {
-        Object* m_as_object;
     };
 
+    RefPtr<Object> m_as_object;
```
Consider Variant?

BenWiederhake replied:
I did, but it seemed like a monstrous approach. Alright, I'll give it another shot.

alimpfard replied:
I'm not sure how big of a change it would be, I just saw a union member being moved out and went "Variant? Variant!".
if it's too annoying, we can totally live with +`sizeof(void*)` there

BenWiederhake commented:
Changes since initial push:
- Rebased for convenience
- Made `Value` inherit from `AK::Variant<>` as Ali suggested

The refactoring was a bit more work than expected, especially since the distinction between `Empty` and `nullptr_t` (which represents the PDF `null` object) really is important. Also, the old `is_int_type<T>` was buggy and always returned `true` if it was any arithmetic type, and the bounds checking was broken.

Please take a moment to appreciate how much work is done by the `using Variant::Variant;` line: It imports *all* constructors from `Variant`, and makes them available as constructors of `Value`.

I'm not entirely sure how to do the `template<IsObject T> Value(RefPtr<T> const& refptr)` case. It's used often enough in LibPDF to warrant being available (instead of just a helper function or something). A different approach I had earlier was:
```
    Value(RefPtr<Object> const& refptr)
        : Variant(refptr ? Variant(NonnullRefPtr<Object>(*refptr)) : Variant(nullptr)) {};
```
However, this is not significantly faster, but harder to read. So I went with Ali's suggestion to initialize it with something empty by default, and then set it to something else if the need arises.

It's a pity that there's no conditional constructor call, something like:
```
    Value(RefPtr<Object> const& refptr)
        : if (refptr) Variant(NonnullRefPtr<Object>(*refptr)) else Variant(nullptr)) {};
```

alimpfard commented on Userland/Libraries/LibPDF/Value.cpp:
```diff
@@ -9,52 +9,31 @@
-    VERIFY_NOT_REACHED();
+    return visit(
+        [&](Empty const&) {
+            // Return type deduction means that we can't use implicit conversions.
```
You can also use a trailing return type for the lambda.

BenWiederhake replied:
Oh, nice! C++ sure has some fancy features :D

alimpfard commented on Userland/Libraries/LibPDF/Value.h:
```diff
@@ -9,171 +9,74 @@
-
-    ~Value() = default;
+    Value(RefPtr<T> const& refptr)
+        : Variant(refptr ? Variant(NonnullRefPtr<Object>(*refptr)) : Variant(nullptr)) {};
```
I'll repeat the "how about assigning nullptr by default and setting it to *refptr in the ctor body" comment.

Also, extraneous semicolon, and please put the braces on a new line.

BenWiederhake replied:
Gah! I did exactly that, but pushed the wrong branch.

@extraneous semicolon: Indeed! Huh, why didn't clang-format pick that one up?

BenWiederhake commented:
Changes since last push:
- Push the correct version of the `Value(RefPtr<>)` constructor. Note that we need to use `Empty` here, do indicate absence of any value, instead of `nullptr`, which would mean the PDF "null" value.
- Use trailing return types for the lambdas.

alimpfard commented:
:thinking: Seeing as "LibPDF: Replace Value class by AK::Variant" essentially undoes "LibPDF: Switch to automatic ref counting, fix memory leak", perhaps they could be squashed to a single commit?
Aside from that, LGTM.

BenWiederhake commented:
> thinking Seeing as "LibPDF: Replace Value class by AK::Variant" essentially undoes "LibPDF: Switch to automatic ref counting, fix memory leak", perhaps they could be squashed to a single commit?

No, since that second-mentioned commit fixes some other ref-counting logic, and I don't want to mix these.

> Aside from that, LGTM.

:partying_face:

alimpfard commented:
:shrug: alright, I guess that's fine.
